site/website/blog/2020-10-06-enabling-full-text-search.md

```diff
@@ -0,0 +1,492 @@
+p.readFrom(mysqlSource)
+ .withoutTimestamps()
+ .map(PetClinicIndexJob::mapChangeRecordToPOJO).setName("mapChangeRecordToPOJO")
+ .mapUsingService(keywordService, PetClinicIndexJob::enrichWithKeywords).setName("enrichWithKeywords")
+ .mapStateful(JoiningState::new, JoiningState::join).setName("JoiningState::join")
+ .writeTo(elasticSink);
+```
+
+Or visualized as a graph:
+
+![Pipeline with global mapping state](assets/2020-10-06-global-join-step.png)
+
+For the pipeline with two step join see [the source
+code](https://github.com/hazelcast-demos/pet-clinic-index-job/blob/two-joins/src/main/java/org/example/jet/petclinic/PetClinicIndexJob.java#L87)
+on Github. This is the graph visualisation of the pipeline:
+
+![Pipeline with 2 separate grouped join steps](assets/2020-10-06-two-join-step.png)
+
+## Testing the job
+
+In order to test the job, we need to create a test database identical to
+the Petclinic database.
+
+We have the following options
+
+* Start the pet clinic application inside the same JVM - the Petclinic
+is a Spring Boot application, running it in-process would be
+relatively easy, but it represents a legacy application, which is
+usually quite difficult to run in such way. Also, there is a high
+probability of dependency clashes, which would make this difficult.
+
+* Start the pet clinic application in a Docker container - it is likely
+that the application is not containerized, we would need to create
+such an image only for the tests.
+
+* Create the database and data using SQL directly - reusing the SQL
+scripts from the Petclinic application and running inserts via JDBC
+seems the easiest way to do this so we will continue with this approach.
+
+Go to
+[Github](https://github.com/hazelcast-demos/pet-clinic-index-job/blob/master/src/main/java/org/example/jet/petclinic/PetClinicIndexJob.java#L121)
+to see the full source code.
+
+## Running the Job
+
+In order to run the indexing job, first checkout the
+[pet-clinic-indexing-job](https://github.com/hazelcast-demos/pet-clinic-index-job/)
+repository
+
+```bash
+git clone https://github.com/hazelcast-demos/pet-clinic-index-job.git
+```
+
+Then build the JAR containing to job
+
+```bash
+cd pet-clinic-indexing-job
+mvn package
+```
+
+And finally submit the job by running the following command:
+
+```bash
+bin/jet submit \
+path/to/pet-clinic-index-job/target/pet-clinic-index-job-1.0-SNAPSHOT-jar-with-dependencies.jar\
+   --database-address localhost \
+   --database-port 3306 \
+   --database-user petclinic \
+   --database-password petclinic \
+   --elastic-host localhost:9200 \
+   --elastic-index petclinic-index
+```
+
+## Petclinic application update
+
+Now that we have the data in Elasticsearch index we can update the
+Petclinic application to use it for search.
+
+There are several changes in the application, but the most important
+changes are Create a search service to search for the data Update the
+search endpoint to use the new search service
```

Comment:
There are several changes in the application, but the most important
changes are:

- Create a search service to search for the data
- Update the search endpoint to use the new search service